workflow/scripts/osemosys_global/powerplant/user_defined_capacity.py

```diff
@@ -0,0 +1,255 @@
+                                 keep='last',
+                                 inplace=True)
+
+        # Add new fuels to FUEL set, if not already present
+        fuel_list = []
+        fuel_list = list(df_iar_custom['FUEL'].unique()) + list(df_oar_custom['FUEL'].unique())
+        fuel_list = list(set(fuel_list))
+        
+        for each_fuel in fuel_list:
+            if each_fuel not in list(fuel_set['VALUE']):
+                fuel_set = pd.concat([fuel_set, pd.DataFrame({'VALUE':[each_fuel]})])          
+        
+        op_life_custom = pd.DataFrame({'TECHNOLOGY': tech_list})
+        
+        for each_tech in op_life_custom['TECHNOLOGY']:
+            op_life_custom.loc[op_life_custom['TECHNOLOGY'] == each_tech,
+                               'VALUE'] = op_life_dict.get(each_tech[3:6])
+
+        op_life_custom['REGION'] = region_name
+        op_life_custom = op_life_custom[['REGION',
+                                         'TECHNOLOGY',
+                                         'VALUE']]
+
+        op_life = pd.concat([op_life_base, op_life_custom])
+        op_life.drop_duplicates(subset=['REGION', 
+                                        'TECHNOLOGY'],
+                                keep='last',
+                                inplace=True)
+
+        # Add CapacityToActivityUnit for custom technologies
+        cap_act_custom = pd.DataFrame({'TECHNOLOGY': tech_list})
+        cap_act_custom.loc[cap_act_custom['TECHNOLOGY'].str.contains('PWR'),
+                           'VALUE'] = 31.536
+        cap_act_custom['REGION'] = region_name
+        cap_act_custom = cap_act_custom[['REGION',
+                                         'TECHNOLOGY',
+                                         'VALUE']]
+        cap_act = pd.concat([cap_act_base, cap_act_custom])
+        cap_act.drop_duplicates(inplace=True)
+        
+        # Update CapitalCost with user-defined costs
+        tech_list = list(tech_capacity_df['TECHNOLOGY'].unique())
+        cap_cost = pd.DataFrame(list(itertools.product(tech_list,
+                                                           get_years(start_year, end_year))),
+                                    columns = ['TECHNOLOGY',
+                                               'YEAR'])
+
+        for each_tech in tech_list:
+            cap_cost.loc[cap_cost['TECHNOLOGY'].str.startswith(each_tech),
+                             'VALUE'] = capex_dict[each_tech]
+
+        cap_cost['REGION'] = region_name
+        cap_cost = cap_cost[['REGION',
+                                     'TECHNOLOGY',
+                                     'YEAR',
+                                     'VALUE']]
+        cap_cost = pd.concat([cap_cost_base, cap_cost])
+        cap_cost.drop_duplicates(subset=['REGION', 'TECHNOLOGY', 'YEAR'],
+                                 keep="last",
+                                 inplace=True)
+        
+        return(df_tech_set, df_max_cap_inv, df_min_cap_inv, 
```

Comment:
Similar to a previous comment; this return type is not guaranteed and can easily lead to an error! If `tech_capacity` is not provided, this function will return `None`. Since you already check for this before calling the function, I would recommend taking out the `if not tech_capacity is None:` at the start of this function, and dedenting this one level - which also improves readability! :)